pkg/server/singleprocess/service_job_test.go

```diff
@@ -672,6 +674,158 @@ func TestServiceGetJobStream_bufferedData(t *testing.T) {
 		},
 	}))
 
+	// Get our job stream and verify we open
+	stream, err := client.GetJobStream(ctx, &pb.GetJobStreamRequest{JobId: queueResp.JobId})
+	require.NoError(err)
+	{
+		resp, err := stream.Recv()
+		require.NoError(err)
+		open, ok := resp.Event.(*pb.GetJobStreamResponse_Open_)
+		require.True(ok, "should be an open")
+		require.NotNil(open)
+
+	}
+
+	// Wait for output, verify its buffered
+	{
+		resp := jobStreamRecv(t, stream, (*pb.GetJobStreamResponse_Terminal_)(nil))
+		event := resp.Event.(*pb.GetJobStreamResponse_Terminal_)
+		require.NotNil(event)
+		require.True(event.Terminal.Buffered)
+		require.Len(event.Terminal.Events, 2)
+	}
+
+	// Send a bit more output now that we're connected
+	require.NoError(runnerStream.Send(&pb.RunnerJobStreamRequest{
+		Event: &pb.RunnerJobStreamRequest_Terminal{
+			Terminal: &pb.GetJobStreamResponse_Terminal{
+				Events: []*pb.GetJobStreamResponse_Terminal_Event{
+					{
+						Event: &pb.GetJobStreamResponse_Terminal_Event_Line_{
+							Line: &pb.GetJobStreamResponse_Terminal_Event_Line{
+								Msg: "hello",
+							},
+						},
+					},
+					{
+						Event: &pb.GetJobStreamResponse_Terminal_Event_Line_{
+							Line: &pb.GetJobStreamResponse_Terminal_Event_Line{
+								Msg: "world",
+							},
+						},
+					},
+				},
+			},
+		},
+	}))
+
+	// Wait for output, verify it's unbuffered (live).
+	{
+		resp := jobStreamRecv(t, stream, (*pb.GetJobStreamResponse_Terminal_)(nil))
+		event := resp.Event.(*pb.GetJobStreamResponse_Terminal_)
+		require.NotNil(event)
+		require.False(event.Terminal.Buffered)
+		require.Len(event.Terminal.Events, 2)
+	}
+
+	// Complete the job
+	require.NoError(runnerStream.Send(&pb.RunnerJobStreamRequest{
+		Event: &pb.RunnerJobStreamRequest_Complete_{
+			Complete: &pb.RunnerJobStreamRequest_Complete{},
+		},
+	}))
+
+	// Should be done
+	_, err = runnerStream.Recv()
+	require.Error(err)
+	require.Equal(io.EOF, err)
+
+	// Wait for completion
+	{
+		resp := jobStreamRecv(t, stream, (*pb.GetJobStreamResponse_Complete_)(nil))
+		event := resp.Event.(*pb.GetJobStreamResponse_Complete_)
+		require.Nil(event.Complete.Error)
+	}
+}
+
+// Tests that a client can connect to a completed job and still read
+// its output.
+// NOTE: this does not test that a client can read job logs for a
+// long-completed job. Some server state implementation (namely this one)
+// do not persist job logs, so streaming completed logs only works
+// if the server hasn't restarted or pruned them from memory.
+func TestServiceGetJobStream_completedBufferedData(t *testing.T) {
```

Comment:
Prior to this, we were only testing reading from a completed job. I renamed that test, and added a new test to read from a job that's still running but has buffered output.